{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7f97fcc8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCCATATACCCA TCCAACTACCCA CCCAGCTACCTA TCCAAGTACCAA ACCAACTACCTA ACCAGATACCGA ACCAGTTACCCA ACCAGTTACCTA ACCAAATACCCA GCCAACTACCAA GCCAAATACCCA TCCAGCTACCCA ACCAAGTACCGA CCCAGATACCGA GCCAAATACCTA ACCATTTACCCA GCCAACTACCTA GCCACATACCCA TCCACATACCCA ACCAATTACCAA ACCATGTACCGA TCCATTTACCTA TCCAGGTACCCA GCCATATACCCA CCCATATACCAA \n"
     ]
    }
   ],
   "source": [
    "\n",
    "f = open(\"dataset_865383_9.txt\",\"r\")\n",
    "\n",
    "k, t = [int(x) for x in next(f).split()]\n",
    "Dna = [str(x) for x in next(f).split()]\n",
    "\n",
    "def GreedyMotifSearch(Dna, k, t):\n",
    "    BestMotifs = []\n",
    "    for i in range(len(Dna)):\n",
    "        BestMotifs.append(Dna[i][:k])\n",
    "    \n",
    "    for m in range(len(Dna[0]) - k + 1):\n",
    "        motifs = [Dna[0][m:m + k]]\n",
    "        for i in range(1,t):\n",
    "            \n",
    "            # Build Profile\n",
    "            Profile = []\n",
    "            for j in range(k):\n",
    "                Profile.append({\"A\":1.0,\"C\":1.0,\"G\":1.0,\"T\":1.0}) # 0.0 for no pseudocounts\n",
    "            for motif in motifs:\n",
    "                for nucleotide in range(len(motif)):\n",
    "                    Profile[nucleotide][motif[nucleotide]] += 1.0\n",
    "            for j in range(k):\n",
    "                for key in Profile[j]:\n",
    "                    Profile[j][key] /= (i + 4) # without adding 4 for no pseudocounts\n",
    "            # End Build Profile\n",
    "            motifs.append(ProfileMostProbableKmer(Dna[i], k, Profile))\n",
    "        \n",
    "        if Score(motifs) < Score(BestMotifs):\n",
    "            BestMotifs = motifs\n",
    "            \n",
    "        \n",
    "    return BestMotifs\n",
    "            \n",
    "            \n",
    "def ProfileMostProbableKmer(Text, k, Profile):\n",
    "    MostProbableKmer = \"\"\n",
    "    KmerProbability = float('-inf')\n",
    "    \n",
    "    for i in range(len(Text) - k + 1):\n",
    "        kmer = Text[i:i + k]\n",
    "        prob = 1\n",
    "        for n in range(len(kmer)):\n",
    "            prob = prob * Profile[n][kmer[n]]\n",
    "            \n",
    "        if prob > KmerProbability:\n",
    "            KmerProbability = prob\n",
    "            MostProbableKmer = kmer\n",
    "    \n",
    "    return MostProbableKmer\n",
    "\n",
    "def Score(motifs):\n",
    "    Profile = []\n",
    "    k = len(motifs[0])\n",
    "    for j in range(k):\n",
    "        Profile.append({\"A\":0.0,\"C\":0.0,\"G\":0.0,\"T\":0.0})\n",
    "    for motif in motifs:\n",
    "        for nucleotide in range(len(motif)):\n",
    "            Profile[nucleotide][motif[nucleotide]] += 1.0\n",
    "    \n",
    "    score = 0\n",
    "    for i in range(k):\n",
    "        max_key = \"A\"\n",
    "        max_val = float('-inf')\n",
    "        for key in Profile[i]:\n",
    "            if Profile[i][key] > max_val:\n",
    "                max_val = Profile[i][key]\n",
    "                max_key = key\n",
    "        \n",
    "        for key in Profile[i]:\n",
    "            if key != max_key:\n",
    "                score += Profile[i][key]\n",
    "                \n",
    "    return score\n",
    "\n",
    "\n",
    "out = \"\"\n",
    "for motif in GreedyMotifSearch(Dna, k, t):\n",
    "    out = out + motif + \" \"\n",
    "print(out)\n",
    "\n",
    "                \n",
    "\n",
    "f.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
