{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b689eefb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['010', '001', '101', '100', '110', '111', '000', '011']\n",
      "defaultdict(<class 'list'>, {'01': ['10', '11'], '00': ['01', '00'], '10': ['01', '00'], '11': ['10', '11']})\n",
      "00011101\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "from copy import deepcopy\n",
    "from random import randint\n",
    "\n",
    "f = open(\"three.txt\",\"r\")\n",
    "\n",
    "k = int(f.readline())\n",
    "\n",
    "def KUniversalString(k):\n",
    "    kmer = \"\"\n",
    "    for i in range(k):\n",
    "        kmer += \"0\"\n",
    "    neighborhood = list(Neighbors(kmer, k))\n",
    "    \n",
    "    return StringReconstruction(neighborhood, k)[k - 1:]\n",
    "\n",
    "def Neighbors(pattern, d):\n",
    "    if d == 0:\n",
    "        return [pattern]\n",
    "    if len(pattern) == 1:\n",
    "        return [\"0\", \"1\"]\n",
    "    \n",
    "    Neighborhood = set()\n",
    "    SuffixNeighbors = Neighbors(pattern[1:], d)\n",
    "    \n",
    "    for string in SuffixNeighbors:\n",
    "        if HammingDistance(string, pattern[1:]) < d:\n",
    "            for nucleotide in [\"0\", \"1\"]:\n",
    "                Neighborhood.add((nucleotide + string))\n",
    "        else:\n",
    "            Neighborhood.add((pattern[:1] + string))\n",
    "    return Neighborhood\n",
    "\n",
    "def HammingDistance(string1, string2):\n",
    "    if len(string1) != len(string2):\n",
    "        return -1\n",
    "    \n",
    "    dist = 0\n",
    "    \n",
    "    for i in range(len(string1)):\n",
    "        if string1[i] != string2[i]:\n",
    "            dist +=1\n",
    "            \n",
    "    return dist\n",
    "\n",
    "def StringReconstruction(Patterns, k):\n",
    "    graph = DeBrujinString(Patterns)\n",
    "    path = EulerianPath(graph)\n",
    "    Text = StringFromGenomePath(path)\n",
    "    return Text\n",
    "    \n",
    "def DeBrujinString(Patterns):\n",
    "    graph = defaultdict(lambda: defaultdict(lambda: 0))\n",
    "    for i in range(len(Patterns)):\n",
    "        graph[Patterns[i][:len(Patterns[i]) - 1]][Patterns[i][1:]] += 1\n",
    "    \n",
    "    reduced_graph = defaultdict(list)\n",
    "    for node1 in graph:\n",
    "        for node2 in graph[node1]:\n",
    "            for i in range(graph[node1][node2]):\n",
    "                reduced_graph[node1].append(node2)\n",
    "    \n",
    "    return reduced_graph\n",
    "\n",
    "def StringFromGenomePath(path):\n",
    "    Text = path[0]\n",
    "    for i in range(1, len(path)):\n",
    "        Text += path[i][-1]\n",
    "        \n",
    "    return Text\n",
    "\n",
    "def EulerianPath(graph): \n",
    "    s = list(graph.keys())[0]\n",
    "    end = FindEnd(graph, s)\n",
    "    graph_r = ReverseGraph(graph)\n",
    "    return FindPath(graph_r, end)\n",
    "\n",
    "def FindEnd(graph, start): \n",
    "    A = deepcopy(graph) #incidence list\n",
    "    \n",
    "    edges_per_vertex = dict()\n",
    "    for key in A:\n",
    "        edges_per_vertex[key] = len(A[key])\n",
    "    \n",
    "    L = [] #stack\n",
    "    K = [] #cycle\n",
    "    \n",
    "    s = start\n",
    "    L.append(s)\n",
    "    v = s\n",
    "    \n",
    "    while len(L) > 0:\n",
    "        if v not in edges_per_vertex:\n",
    "            return v\n",
    "        \n",
    "        if edges_per_vertex[v] > 0:\n",
    "            i = randint(0, len(A[v]) - 1)\n",
    "            L.append(v)\n",
    "            u = A[v][i]\n",
    "            edges_per_vertex[v] -= 1\n",
    "            A[v].pop(i)\n",
    "            v = u\n",
    "        else:\n",
    "            K.append(v)\n",
    "            v = L[-1]\n",
    "            L.pop()\n",
    "    \n",
    "    return K[0]\n",
    "        \n",
    "def FindPath(graph, end):\n",
    "    A = deepcopy(graph) #incidence list\n",
    "    \n",
    "    edges_per_vertex = dict()\n",
    "    for key in A:\n",
    "        edges_per_vertex[key] = len(A[key])\n",
    "    \n",
    "    L = [] #stack\n",
    "    K = [] #cycle\n",
    "    \n",
    "    s = end\n",
    "    L.append(s)\n",
    "    v = s\n",
    "    start_flag = 0\n",
    "    \n",
    "    while len(L) > 0:\n",
    "        \n",
    "        if v not in edges_per_vertex:\n",
    "            start_flag = 1\n",
    "            start = v\n",
    "            A[v].append(end)\n",
    "            edges_per_vertex[v] = 0\n",
    "            edges_per_vertex[v] += 1\n",
    "        \n",
    "        if edges_per_vertex[v] > 0:\n",
    "            L.append(v)\n",
    "            u = A[v][-1]\n",
    "            edges_per_vertex[v] -= 1\n",
    "            A[v].pop()\n",
    "            v = u\n",
    "        else:\n",
    "            K.append(v)\n",
    "            v = L[-1]\n",
    "            L.pop()\n",
    "    \n",
    "    return K\n",
    "    \n",
    "\n",
    "def ReverseGraph(graph):\n",
    "    graph_r = defaultdict(list)\n",
    "    for key in graph:\n",
    "        for i in graph[key]:\n",
    "            graph_r[i].append(key)\n",
    "            \n",
    "    return graph_r\n",
    "\n",
    "print(KUniversalString(k))\n",
    "\n",
    "f.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
